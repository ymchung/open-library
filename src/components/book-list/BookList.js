import React, { useEffect } from 'react';
import PropTypes, { arrayOf, string, shape, number, bool, func } from 'prop-types';
import { useSearchParams, useLocation, useNavigate } from 'react-router-dom';

import Pagination from '../pagination';
import Loading from '../loading';
import BookCard from '../book-card';


const namespace = 'book-list';

const Books = ({ books, searchText, isLoading, fetchBooks }) => {
    const [searchParams, setSearchParams] = useSearchParams();
    const location = useLocation();
    const navigate = useNavigate();
    const bookParam = searchParams.get('q');

    useEffect(() => {
        if (bookParam) {
            const pageParam = searchParams.get('page') || 1;
            fetchBooks(bookParam, pageParam);
        }
    }, [bookParam]);

    const handleOnClickPage = selectedPage => {
        fetchBooks(searchText, selectedPage);
        const params = new URLSearchParams(location.search);
        const objParams = {};
        for (let p of params) {
            objParams[p[0]] = p[1];
          }
        setSearchParams({...objParams, page: selectedPage});
    }

    const handleOnBook = (e, k) => {
        e.preventDefault();
        const bookKey = k.replace(new RegExp('.*' + '/works/'), '');
        navigate(`/book/${bookKey}`);
    };

    return (
        <div className={`${namespace} row`}>
            {isLoading && <Loading />}
            {books && books.docs && !!books.docs.length && books.docs.map(book => 
                <BookCard 
                    key={book.key}
                    book={book.key}
                    title={book.title}
                    authorName={book.author_name}
                    year={book.first_publish_year}
                    handleOnBook={handleOnBook}
                />)}
            {!isLoading && books && books.booksNumFound === 0 && <p>Lo sentimos, no encontramos resultados</p>}
            {books && <Pagination currentPage={books.currentPage} total={books.booksNumFound} pageLimit={100} onPageChange={handleOnClickPage} />}
        </div>
    );  
};

Books.propTypes = {
    books: shape({
        docs: arrayOf(shape({
            key: string,
            title: string,
            first_publish_year: number,
            author_name: string,
        })),
        currentPage: number,
        booksNumFound: number,
    }),
    isLoading: bool,
    searchText: string,
    fetchBooks: func.isRequired,
};

Books.defaultProps = {
    books: null,
    isLoading: false,
    searchText: null,
};

export default Books;