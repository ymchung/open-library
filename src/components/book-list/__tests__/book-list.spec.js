import { render } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from "history";
import BookList from '../BookList';

describe('BookList Component', () => {
    const props = {
        books: {
            docs: [
                {
                    key: 'A123B',
                    title: 'Harry Potter',
                    first_publish_year: 1999,
                    author_name: 'J. K. Rowling',
                },
                {
                    key: 'A321C',
                    title: 'Harry Potter 2',
                    first_publish_year: 2000,
                    author_name: 'J. K. Rowling',
                },
            ],
            currentPage: 1,
            booksNumFound: 2,
        },
        isLoading: true,
        searchText: 'Harry',
        fetchBooks: jest.fn(),
    }

    test('Component should render correctly', () => {
        const history = createMemoryHistory();
        const { container } = render(
            <Router location={history.location}>
                <BookList {...props} />
            </Router>
        );
        expect(container.firstChild).toMatchSnapshot();
    });
});