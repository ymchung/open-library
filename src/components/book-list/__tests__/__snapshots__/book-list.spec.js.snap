// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookList Component Component should render correctly 1`] = `
<div
  class="book-list row"
>
  <div
    class="loading"
  >
    <div
      class="spinner-border"
      role="status"
    >
      <span
        class="visually-hidden"
      >
        Cargando...
      </span>
    </div>
  </div>
  <div
    class="book-card card"
  >
    <div
      class="card-body"
    >
      <a
        class="book-card__link"
        data-testid="book-card__link"
        href="/#"
      >
        <h3
          class="book-card__title"
        >
          Harry Potter
        </h3>
        <p
          class="book-card__subtitle"
        >
          de 
          J. K. Rowling
        </p>
        <p
          class="book-card__text"
        >
          Publicado por primera vez en 
          1999
        </p>
      </a>
    </div>
  </div>
  <div
    class="book-card card"
  >
    <div
      class="card-body"
    >
      <a
        class="book-card__link"
        data-testid="book-card__link"
        href="/#"
      >
        <h3
          class="book-card__title"
        >
          Harry Potter 2
        </h3>
        <p
          class="book-card__subtitle"
        >
          de 
          J. K. Rowling
        </p>
        <p
          class="book-card__text"
        >
          Publicado por primera vez en 
          2000
        </p>
      </a>
    </div>
  </div>
</div>
`;
